### ISSUE TITLE
Issue with links in config.json that have the same "name"

### ISSUE THREAD

**Issue by grezniczek** [vanderbilt/redcap-external-modules#310]
Link processing relies on "name" to be unique, which prevents two links to have the same "name". Only one will get displayed. Any same-named are dropped.

There may be reasons two have two links with the same name, but different icons and urls. I had one (it could be argued if it was a good one) and that's why I noticed this.

Could we have a "key" property to keep track of links, like elsewhere (settings)? Since "name" is what actually is displayed and could be subject to translation, having a "key" would be very much preferable for identfying links (e.g. in the hook for processing). This would be fully backward / forward compatible, too, as when no "key" is in defined in config.json, we can use the value of "name" as fall back, and existing logic relies on "name" anyways (or on using their own sneaky addition to config.json - I'd rather use a documented "key", though).

While at this, it would be nice to be able to specify "javascript: ..." in the url (there are ways to do it with the link hook now, but it's a hack).

@mmcev106 -  if you approve of the idea, you can assign me this issue and I'll make a PR for this (this should be quite easy).

**Comment by mmcev106** [comment]
The links feature definitely needs help maturing.  I'm having troubling thinking of a use case that's considered best practice where it makes sense to have two links with the same name......perhaps if we create a way to add links programmatically?  I wonder if we should wait until we have a better use case for introducing a "key", so that we can make sure our implementation fits that case well.

**Comment by grezniczek** [comment]
Translation is a use case already. It breaks referring to a link, e.g. in the callback, by hardcoded name. tt_name could serve that, but I'd rather have "key", as "key" is established in many other cases: system/project-settings, the upcoming tabs (yes, I started working on the config dialog replacement ;), ..

**Comment by mmcev106** [comment]
A link's `url` could effectively be considered a key already, since it must be unique.  I've used `url` checking for a few things in `redcap_module_link_check_display()` before.  Is there a benefit to adding another unique indentifier?

**Comment by grezniczek** [comment]
Readability in code? `my-link` versus `https://some.deep.d9sdffssd.subdomain.service.tld/a/b/c/d?e=A1587740673487&ei=AQCjXq-yHYiWkwXb1bDADQ&thats=enough_now#ornot`.
Multiply by 10. 
You have to maintain code **and** config.json.

Plus: the key could be inserted into the link html (data-module-link="prefix-key") for jQuery goodness later on.

**Comment by mmcev106** [comment]
I was thinking of using the `url` specified in config.json, which is typically something like `my-link`.  Are we thinking about different use cases?

**Comment by grezniczek** [comment]
Ah, sorry. I had a module where I wanted an external link ... so I did modify it in the callback hook (and had used a dummy value in url, and was already thinking ahead to allowing more options for url in config.json.

But consider this: You have a folder structure inside your module directory, and you want to link deep. That gets you something like "folder/level/three/file.php". Now, in PHP, you have to have that whole string to match, or need to do some subpart matching.
Now, you want to reorganize your module's folder structure down the road ... and file.php ends up in another folder that consolidates many and is now called file2.php because there was another file.php. At this point you can only hope you used the full path in PHP so you can do a simple search/replace.

**Comment by mmcev106** [comment]
At the end of the day, I'd love to see what you end up with on a PR here.  I'm open to the idea of introducing link keys, but want to make sure it's right long term path since we've gotten into trouble in the past with over-engineering for less common situations, and ending up with something that didn't serve the common situations well.

**Comment by grezniczek** [comment]
:)
I'll come up with one soon.

**Comment by mmcev106** [comment]
Here's my thought process, in case it's helpful:
- If we add link keys, we'll have to grandfather in existing key-less links
- This makes we wonder if link keys should be optional long term since most projects we have today don't seem to need them.
- Which makes we wonder if we need them in the first place...
- I suppose they could default to the `url`, and overriding them could be optional.  I wonder how that idea lines up with possible implementations...

**Comment by grezniczek** [comment]
I don't think we have to worry about this at all. The callback gets the whole link "object". Up to module authors what they do with that, and it contains whatever they put into config.json. 

The only change is internal (not using "name" but a guaranteed unique key/id), and in documentation, which will recommend to use the "key" and that it must be unique (within the "links" context) when used.

**Comment by grezniczek** [comment]
I have just submitted a PR for this #317.

I noticed and addressed some other issues as well.

The naming collison reported originally affected also links of different modules! So same-named links of different modules would overwrite each other (last one wins).
This is now fixed by prepending the array key used for internal storing of links with the unique prefix.

External and Javascript links are now supported. The original docs mentioned external, but as far as I could see, this never worked.

**Comment by grezniczek** [comment]
One could think about adding little headers (the module name) into the HTML output of links, so to make clear to the user which links belong to which module (as this may not be readily apparent from the name).

**Comment by grezniczek** [comment]
To stress again: These changes are fully transparent to existing module code.
